examples/platform/silabs/efr32/rs911x/hal/efx_spi.c

```diff
@@ -141,13 +158,131 @@ void rsi_hal_board_init(void)
     spiTransferLock = xSemaphoreCreateBinaryStatic(&xEfxSpiIntfSemaBuffer);
     xSemaphoreGive(spiTransferLock);
 
+#if defined(EFR32MG24)
+    spi_sem_sync_hdl = xSemaphoreCreateBinaryStatic(&spi_sem_peripheral);
+    xSemaphoreGive(spi_sem_sync_hdl);
+#endif /* EFR32MG24 */
+
     /* GPIO INIT of MG12 & MG24 : Reset, Wakeup, Interrupt */
     sl_wfx_host_gpio_init();
 
     /* Reset of Wifi chip */
     sl_wfx_host_reset_chip();
+    SILABS_LOG("%s completed.", __func__);
+}
+
+#if defined(EFR32MG24)
+
+void SPIDRV_ReInit(uint32_t baudrate)
+{
+    if (EUSART_BaudrateGet(MY_USART) == baudrate)
+    {
+        // EUSART synced to baudrate already
+        return;
+    }
+    EUSART_BaudrateSet(MY_USART, 0, baudrate);
+}
+
+sl_status_t sl_wfx_host_spi_cs_assert(void)
+{
+    configASSERT(spi_sem_sync_hdl);
+    if (xSemaphoreTake(spi_sem_sync_hdl, portMAX_DELAY) != pdTRUE)
+    {
+        return SL_STATUS_TIMEOUT;
+    }
+
+    if (!spi_enabled) // Reduce SPIDRV_Init
+    {
+        SPIDRV_Init(SL_SPIDRV_HANDLE, &sl_spidrv_eusart_init_exp);
+        spi_enabled = true;
+    }
+    GPIO_PinOutClear(SL_SPIDRV_EUSART_EXP_CS_PORT, SL_SPIDRV_EUSART_EXP_CS_PIN);
+    return SL_STATUS_OK;
+}
+
+sl_status_t sl_wfx_host_spi_cs_deassert(void)
+{
+    if (spi_enabled)
+    {
+        SPIDRV_DeInit(SL_SPIDRV_HANDLE);
+        spi_enabled = false;
+    }
+    GPIO_PinOutSet(SL_SPIDRV_EUSART_EXP_CS_PORT, SL_SPIDRV_EUSART_EXP_CS_PIN);
+    GPIO->EUSARTROUTE[SL_SPIDRV_EUSART_EXP_PERIPHERAL_NO].ROUTEEN = 0;
+    xSemaphoreGive(spi_sem_sync_hdl);
+    return SL_STATUS_OK;
+}
+
+void sl_wfx_host_spiflash_cs_assert(void)
+{
+    GPIO_PinOutClear(SL_MX25_FLASH_SHUTDOWN_CS_PORT, SL_MX25_FLASH_SHUTDOWN_CS_PIN);
 }
 
+void sl_wfx_host_spiflash_cs_deassert(void)
+{
+    GPIO_PinOutSet(SL_MX25_FLASH_SHUTDOWN_CS_PORT, SL_MX25_FLASH_SHUTDOWN_CS_PIN);
+}
+
+void sl_wfx_host_pre_bootloader_spi_transfer(void)
+{
+    if (xSemaphoreTake(spi_sem_sync_hdl, portMAX_DELAY) != pdTRUE)
+    {
+        return;
+    }
+    if (spi_enabled)
+    {
+        SPIDRV_DeInit(SL_SPIDRV_HANDLE);
+        spi_enabled = false;
+    }
+    int32_t status = bootloader_init();
+    // bootloader_init takes care of SPIDRV_Init()
+    if (status != BOOTLOADER_OK)
+    {
+        SILABS_LOG("%s: bootloader_init failed with error code: %d", __func__, status);
+        return;
```

Comment:
This function seems crucial .... Should have a return value that will be check by the code instead of simply logging a low-level error in the logs. Also to prevent further stack execution should this function fails.